warehouse/tracker.go

```diff
@@ -0,0 +1,183 @@
+		query         string
+		queryArgs     []interface{}
+		createdAt     sql.NullTime
+		exists        bool
+		uploaded      int
+		syncFrequency = "1440"
+		Now           = timeutil.Now
+		NowSQL        = "NOW()"
+		timeWindow    = config.GetInt("Warehouse.uploadBufferTimeInMin", 180)
+		source        = warehouse.Source
+		destination   = warehouse.Destination
+	)
+
+	if wh.NowSQL != "" {
+		NowSQL = wh.NowSQL
+	}
+	if wh.Now != nil {
+		Now = wh.Now
+	}
+
+	if !source.Enabled || !destination.Enabled {
+		return nil
+	}
+
+	excludeWindow := warehouseutils.GetConfigValueAsMap(warehouseutils.ExcludeWindow, warehouse.Destination.Config)
+	excludeWindowStartTime, excludeWindowEndTime := GetExcludeWindowStartEndTimes(excludeWindow)
+	if CheckCurrentTimeExistsInExcludeWindow(Now(), excludeWindowStartTime, excludeWindowEndTime) {
+		return nil
+	}
+
+	if sf := warehouseutils.GetConfigValue(warehouseutils.SyncFrequency, *warehouse); sf != "" {
+		syncFrequency = sf
+	}
+	if value, err := strconv.Atoi(syncFrequency); err == nil {
+		timeWindow += value
+	}
+
+	query = fmt.Sprintf(`
+				SELECT
+				  created_at
+				FROM
+				  %[1]s
+				WHERE
+				  source_id = '%[2]s' AND
+				  destination_id = '%[3]s' AND
+				  created_at > %[6]s - interval '%[4]d MIN' AND
+				  created_at < %[6]s - interval '%[5]d MIN'
+				ORDER BY
+				  created_at DESC
+				LIMIT
+				  1;
+				`,
+		warehouseutils.WarehouseStagingFilesTable,
+		source.ID,
+		destination.ID,
+		2*timeWindow,
+		timeWindow,
+		NowSQL,
+	)
+
+	err := wh.dbHandle.QueryRowContext(ctx, query).Scan(&createdAt)
+	if err == sql.ErrNoRows {
+		return nil
+	}
+	if err != nil && err != sql.ErrNoRows {
+		return fmt.Errorf("fetching last upload time for source: %s and destination: %s: %w", source.ID, destination.ID, err)
+	}
+
+	if !createdAt.Valid {
+		return nil
+	}
+
+	query = fmt.Sprintf(`
+				SELECT
+				  EXISTS (
+					SELECT
+					  1
+					FROM
+					  %s
+					WHERE
+					  source_id = $1 AND
+					  destination_id = $2 AND
+					  (
+						status = $3
+						OR status = $4
+						OR status LIKE $5
+					  ) AND
+					  updated_at > $6
+				  );
+	`,
+		warehouseutils.WarehouseUploadsTable,
+	)
+	queryArgs = []interface{}{
+		source.ID,
+		destination.ID,
+		model.ExportedData,
+		model.Aborted,
+		"%_failed",
+		createdAt.Time.Format(misc.RFC3339Milli),
+	}
+
+	err = wh.dbHandle.QueryRowContext(ctx, query, queryArgs...).Scan(&exists)
+	if err != nil && err != sql.ErrNoRows {
+		return fmt.Errorf("fetching last upload status for source: %s and destination: %s: %w", source.ID, destination.ID, err)
+	}
+
+	if exists {
+		uploaded = 1
+	}
+
+	tags := stats.Tags{
+		"workspaceId": warehouse.WorkspaceID,
+		"module":      moduleName,
+		"destType":    wh.destType,
+		"warehouseID": misc.GetTagName(
+			destination.ID,
+			source.Name,
+			destination.Name,
+			misc.TailTruncateStr(source.ID, 6)),
+	}
+	wh.stats.NewTaggedStat("warehouse_successful_upload_exists", stats.CountType, tags).Count(uploaded)
```

Comment:
I would suggest that instead of reporting 0 or 1, report:

max( `now() - max(created_at, excludeWindowEndTime)` of each upload)

We can later set up the rule based on sync time

Reply:
@achettyiitr, what do you think of this? Can we add a metric to capture the max, avg, and sum waiting time?

Reply:
We will be tracking this as a separate task?